purchases/src/main/kotlin/com/revenuecat/purchases/galaxy/handler/ProductDataHandler.kt

```diff
@@ -0,0 +1,182 @@
+import com.revenuecat.purchases.galaxy.GalaxyStrings
+import com.revenuecat.purchases.galaxy.IAPHelperProvider
+import com.revenuecat.purchases.galaxy.listener.ProductDataResponseListener
+import com.revenuecat.purchases.galaxy.toStoreProduct
+import com.revenuecat.purchases.galaxy.utils.isError
+import com.revenuecat.purchases.models.StoreProduct
+import com.samsung.android.sdk.iap.lib.vo.ErrorVo
+import com.samsung.android.sdk.iap.lib.vo.ProductVo
+import java.util.ArrayList
+
+internal class ProductDataHandler(
+    private val iapHelper: IAPHelperProvider,
+    private val mainHandler: Handler,
+) : ProductDataResponseListener {
+
+    companion object {
+        private const val GET_PRODUCT_DATA_TIMEOUT_MILLIS = 10_000L
+    }
+
+    @get:Synchronized
+    private var inFlightRequest: Request? = null
+
+    private data class Request(
+        val productIds: Set<String>,
+        val productType: ProductType,
+        val onReceive: StoreProductsCallback,
+        val onError: PurchasesErrorCallback,
+    )
+
+    @get:Synchronized
+    internal val productDataCache = mutableMapOf<String, ProductVo>()
+
+    override fun getProductDetails(
+        productIds: Set<String>,
+        productType: ProductType,
+        onReceive: (List<StoreProduct>) -> Unit,
+        onError: (PurchasesError) -> Unit,
+    ) {
+        if (productIds.isEmpty()) {
+            // Note: This early exit is more than just an optimization. If we call iapHelper.getProductsDetails
+            // with an empty string param (this would happen if productIds is empty), then iapHelper.getProductsDetails
+            // returns all products for the app, which wouldn't give us the product we want (none).
+            log(LogIntent.DEBUG) { GalaxyStrings.EMPTY_GET_PRODUCT_DETAILS_REQUEST }
+            onReceive(emptyList())
+            return
+        }
+
+        if (inFlightRequest != null) {
+            log(LogIntent.GALAXY_ERROR) { GalaxyStrings.ANOTHER_GET_PRODUCT_DETAILS_REQUEST_IN_FLIGHT }
+            val error = PurchasesError(
+                code = PurchasesErrorCode.OperationAlreadyInProgressError,
+                underlyingErrorMessage = "Only one Galaxy Store product request is allowed at a time.",
+            )
+            onError(error)
+            return
+        }
+
+        log(LogIntent.DEBUG) { GalaxyStrings.REQUESTING_PRODUCTS.format(productIds.joinToString()) }
+
+        synchronized(lock = this) { productDataCache.toMap() }.let { productDataCache ->
+            val request = Request(
+                productIds = productIds,
+                productType = productType,
+                onReceive = onReceive,
+                onError = onError,
+            )
+
+            if (productDataCache.keys.containsAll(productIds)) {
+                val cachedProducts = productIds.mapNotNull(productDataCache::get)
+
+                synchronized(this) {
+                    this.inFlightRequest = request
```

Comment:
I guess there is a small chance of a race condition between this being set, and when we check at the beginning of this method. I wonder if we should maybe only check whether there is an inflight request here?

Reply:
Might be worth removing all the synchronous code, and instead create a single thread to execute all these Galaxy operations. We've created single thread executors for backend requests, and we could follow a similar approach. See: https://github.com/RevenueCat/purchases-android/blob/123f0bc295042a816a73c2edf3382791837afbc6/purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesFactory.kt#L485